hamilton/function_modifiers/expanders.py

```diff
@@ -869,6 +874,156 @@ def extractor_fn(
+        ):
+            field_name = fields[idx]
+            errors.append(f"Field {field_name} (index {idx}) does not declare a valid type: {arg}")
+        field_types.append(arg)
+
+    if errors:
+        raise base.InvalidDecoratorException(f"Found errors in the output type: {errors}")
+
+    return field_types
+
+
+class unpack_fields(base.SingleNodeNodeTransformer):
+    """Extracts fields from a tuple output.
+
+    Expands a single function into the following nodes:
+
+    - 1 function that outputs the original tuple
+    - n functions, each of which take in the original tuple and output a specific field
+
+    The decorated function must have an return type of either `tuple` (python 3.9+) or
+    `typing.Tuple`, and must specify either:
+    - An explicit length tuple (e.g.`tuple[int, str]`, `typing.Tuple[int, str]`)
+    - An indeterminate length tuple (e.g. `tuple[int, ...]`, `typing.Tuple[int, ...]`)
+
+    :param fields: Fields to extract from the return value of the decorated function.
+    """
+
+    output_type: Any
+    field_types: List[Type]
+
+    def __init__(self, *fields: str):
+        super().__init__()
+        self.fields = list(fields)
+
+    @override
+    def validate(self, fn: Callable):
+        output_type = typing.get_type_hints(fn).get("return")
+        field_types = _process_unpack_fields(self.fields, output_type)
+        if field_types:
+            self.field_types = field_types
+            self.output_type = output_type
+        else:
+            message = (
+                f"For unpacking fields, the decorated function output type must be either an "
+                f"explicit length tuple (e.g.`tuple[int, str]`, `typing.Tuple[int, str]`) or an "
+                f"indeterminate length tuple (e.g. `tuple[int, ...]`, `typing.Tuple[int, ...]`), "
+                f"not: {output_type}"
+            )
+            raise base.InvalidDecoratorException(message)
+
+    @override
+    def transform_node(
+        self, node_: node.Node, config: Dict[str, Any], fn: Callable
+    ) -> Collection[node.Node]:
+        fn = node_.callable
+        base_doc = node_.documentation
+        base_tags = node_.tags.copy()
+
+        if inspect.iscoroutinefunction(fn):
+
+            async def tuple_generator(*args, **kwargs):  # type: ignore
+                tuple_generated = await fn(*args, **kwargs)
+                return tuple_generated
+
+        else:
+
+            def tuple_generator(*args, **kwargs):
+                tuple_generated = fn(*args, **kwargs)
+                return tuple_generated
+
+        output_nodes = [node_.copy_with(callabl=tuple_generator)]
+
+        for idx, (field_name, field_type) in enumerate(zip(self.fields, self.field_types)):
+            # NOTE: The extractors, as defined below, are constructed to avoid closure issues.
+            if inspect.iscoroutinefunction(fn):
+
+                async def extractor(field_index: int = idx, **kwargs) -> field_type:  # type: ignore
```

Comment:
Is there any reason to have an async extractor? It'll be the same either way as `await` is never called...

Reply:
To handle this case, yeah?
```python

@unpack_fields
async def my_function(...) -> Tuple[..., int]:
   ...
 ```
 Otherwise all async functions should have an `await` in them though right?

Reply:
The extractor function goes after it, it shouldn't need to be async, as it doesn't do anything asynchronous. The function itself is async (which forms a node), but the extractor doesn't need to be.

Reply:
Ah, ya got me, I was copying parts of `extract_fields` and didn't notice that there was no `await` in the extractor. I don't think we need it considering (as @elijahbenizzy said) it's only indexing the result of `tuple_generator` (either sync or async). Commit incoming.